### PR TITLE
Add provider free2move / free2move_stuttgart

### PR DESCRIPTION
**Pull request by hbruch**
This PR adds `Free2moveProvider`. 

NOTE:  the generated feed publishes free floating vehicles using the original vehicle identifiers (`VIN`) as `bike_id`s and in `rental_uri`s. In consequence, it does not comply to the GBFS spec, which requires that

>  The bike_id identifier MUST be rotated to a random string after each trip to protect user privacy (as of v2.0). Use of persistent vehicle IDs poses a threat to user privacy. The bike_id identifier SHOULD only be rotated once per trip.

This feed must not be published openly without handling of this issue.

### CODE REVIEW

**Review by derhuerst — CHANGES_REQUESTED** [review]
I'd like to tweak the `geofencing_zones.json` docs and the caching implementation. Otherwise just small comments.

File: docs/mappings/free2move_gbfs_2.3_mapping.md [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+}
+```
+
+
+#### Field Mappings
+
+GBFS Field | Mapping
+--- | ---
+`vehicle_type_id` |  `row['buildSeries']` appended with the vehicle's primaryColor.
+`form_factor` | Always `car`.
```
[the remark this answers is not in the document]
Comment (hbruch) [reply docs/mappings/free2move_gbfs_2.3_mapping.md:112]: There is no vehicle type in their API, all buildSeries are cars, so we need to assume it. I don't see any way to assert it

File: x2gbfs/providers/free2move.py [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+        else:
+            cached_vehicles_response = {}
+            parameters = None
+            refresh_base_file = True
+
+        vehicles_response = self._get_with_authorization(
+            self.VEHICLES_URL_TEMPLATE.format(location_alias=location_alias), params=parameters
+        )
+        vehicles_delta_file_path = cachedir_for_location / f'vehicles_{location_alias}_delta.json'
+        vehicles_delta_file_path.write_text(json.dumps(vehicles_response))
```
[the remark this answers is not in the document]
Comment (hbruch) [reply x2gbfs/providers/free2move.py:132]: This is intended for debugging purposes.

File: x2gbfs/providers/free2move.py [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+        """
+        Merge cached responses with delta responses. Merge is performed on VIN (vehicle ID number):
+        If delta information is available for a vehicle, it is used, otherwise the cached information.
+        Note: that way, vehicles which are no longer available and not contained in delta updates
+        might never receive an update. As consequence a full update should be performed every N hours...
+        """
+        # We build dicts with VINs as key to merge on them
+        cached_vehicles_map = {v['vin']: v for v in cached_response.get('vehicles', {})}
+        delta_vehicles_map = {v['vin']: v for v in delta_response.get('vehicles', {})}
+        merged_vehicles_map = cached_vehicles_map | delta_vehicles_map
```
[the remark this answers is not in the document]
Comment (hbruch) [reply x2gbfs/providers/free2move.py:153]: The API publishes `freeForRental`, so unavailable vehicles should be published in the delta. Using only delta feeds nvertheless could result in "ghost cars", which are not published any longer. Performing a full update every hour should limit this at least.

File: x2gbfs/util.py [hunk 4]
```diff
@@ -38,3 +41,27 @@ def get(
...
     return response
+
+
+def post(
+    url: str,
+    params: Optional[dict[str, str]] = None,
+    headers: Optional[dict[str, str]] = None,
+    json: Optional[Dict] = None,
+    timeout: int = 5,
+    user_agent: str = 'x2gbfs +https://github.com/mobidata-bw/',
```
[the remark this answers is not in the document]
Comment (hbruch) [reply x2gbfs/util.py:52]: Yes, this should become a proper PR

File: x2gbfs/providers/free2move.py [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+        'PEUGEOT_2008': ['Peugeot', '2008'],
+        'PEUGEOT_308': ['Peugeot', '308'],
+        'PEUGEOT_3008': ['Peugeot', '3008'],
+        'RENAULT_ZOE': ['Renault', 'ZOE'],
+        'OPEL_ASTRA': ['Opel', 'Astra'],
+        'OPEL_CORSA': ['Opel', 'Corsa'],
+        'OPEL_CROSSLAND': ['Opel', 'Crossland'],
+    }
+
+    COLOR_MAPPING = {
```
[the remark this answers is not in the document]
Comment (hbruch) [reply x2gbfs/providers/free2move.py:251]: There is no such naming schema, the given values are listed in the (non-public) documentation

File: docs/mappings/free2move_gbfs_2.3_mapping.md [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+### system_alerts.json
+
+No endpoint.
+
+### geofencing_zones.json
+
+Geofencing zones are extracted from the `operation_area` endpoint, which returns a single GeoJSON feature.
+
+The geofencing zones are generated as a one element collection with a MultiPolygon counterclockwise ordered lists of coordinates (to reflect the restrictions outside the service area) and the following properties:
```
[the remark this answers is not in the document]
Comment (hbruch) [reply docs/mappings/free2move_gbfs_2.3_mapping.md:251]: It is a MulitPolygon

File: x2gbfs/providers/free2move.py [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+    def all_vehicles(self, location_alias) -> Generator[dict, None, None]:
+        """
+        Requests all vehicles and returns a Generator to iterate over.
+        """
+        cachedir_for_location = self.cachedir / f'free2move_{location_alias}'
+        if not cachedir_for_location.is_dir():
+            cachedir_for_location.mkdir(parents=True, exist_ok=True)
+
+        # if cached_file exists and newer than, load delta and merge
+        vehicles_base_file_path = cachedir_for_location / f'vehicles_{location_alias}_base.json'
```
[the remark this answers is not in the document]
Comment (hbruch) [reply x2gbfs/providers/free2move.py:110]: I decided against, as one API instance may be used with multiple locations. the cachedir itself is assigned in `__init__()`

File: x2gbfs/providers/free2move.py [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+        feature_collection = get(self.CHARGING_POINT_URL_TEMPLATE.format(location_alias=location_alias)).json()
+        for feature in feature_collection.get('features', {}):
+            yield feature
+
+    def all_vehicles(self, location_alias) -> Generator[dict, None, None]:
+        """
+        Requests all vehicles and returns a Generator to iterate over.
+        """
+        cachedir_for_location = self.cachedir / f'free2move_{location_alias}'
+        if not cachedir_for_location.is_dir():
```
[the remark this answers is not in the document]
Comment (hbruch) [reply x2gbfs/providers/free2move.py:106]: I decided against, as one API instance may be used with multiple locations. the cachedir itself is assigned in __init__()

File: x2gbfs/providers/free2move.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (derhuerst) [reply x2gbfs/providers/free2move.py:153]: > Using only delta feeds nvertheless could result in "ghost cars", which are not published any longer. Performing a full update every hour should limit this at least.

I recommend adding this information to the docs.

File: docs/mappings/free2move_gbfs_2.3_mapping.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (derhuerst) [reply docs/mappings/free2move_gbfs_2.3_mapping.md:251]: So there is a single `MultiPolygon` with a single polygon inside, which resembles a "limitation area" (we explicitly reverse its coordinates, after all)?

File: x2gbfs/util.py [hunk 9]
```diff
@@ -38,3 +41,27 @@ def get(
...
+    request_headers = dict(headers) if headers is not None else {}
+    request_headers['User-Agent'] = user_agent
+    response = requests.post(url, headers=request_headers, timeout=timeout, params=params, json=json)
+    response.raise_for_status()
+    return response
+
+
+def reverse_multipolygon(geometry):
+    if geometry and geometry.get('type') == 'MultiPolygon':
+        for single_polygon in geometry.get('coordinates', []):
```
Comment (derhuerst) [inline x2gbfs/util.py:63]: Shall we assert – or at least document the assumption – that it's actually just *one* polygon?

Comment (hbruch) [reply x2gbfs/util.py:63]: The code asserts that it's a MultiPolygon and iterates over all it's polygons. Maybe  `single_polygon` (for one of the multiple polygons) is missleading you?

**Review by ThorstenFroehlinghaus — APPROVED** [review]
This converter works for all locations, right? I need to clarify, if we can provide other locations than stuttgart.

File: x2gbfs/util.py (hunk 9 above)
Comment (derhuerst) [reply x2gbfs/util.py:63]: If there are multiple polygons, is there a non-limited service area, whose coordinates *should not* be reversed?

File: docs/mappings/free2move_gbfs_2.3_mapping.md [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+`bike_id` | `row['vid']` (NOTE: the [VIN](https://en.wikipedia.org/wiki/Vehicle_identification_number) is not rotated by Free2Move. As a consequence, the generated GBFS feed *should not* be published publicly to be GDPR-compliant. It is, in consequence, not appropriate for unrestriced use. Consumers must handle GDPR requirements.)
+`lat` | - `row['geoCoordinate']['latitude']`
+`lon` | - `row['geoCoordinate']['longitude']`
+`is_reserved` | `not row['freeForRental']`
+`is_disabled` | `False`
+`rental_uris` | -
+`vehicle_type_id` | see section vehicle_types
+`last_reported` | curent time
+`current_range_meters` | `row['remainingRange']`, if set
+`current_fuel_percent` | `row['fuelLevel'] / 100.0`, or `.25` if unset
```
[the remark this answers is not in the document]
Comment (hbruch) [reply docs/mappings/free2move_gbfs_2.3_mapping.md:227]: According to documentation, `fuelLevel` is not optional and should provided for every vehicle. I remove the default handling.

File: x2gbfs/util.py (hunk 9 above)
Comment (hbruch) [reply x2gbfs/util.py:63]: No, there are not. GBFSv2 uses a non-spec-conformant GeoJSON style to address areas surrounding the operational area. I think we can't do much about it in GBFSv2 but try to migrate to GBFSv3, which has a much cleaner way of handling out-of-service-area areas. (using standard GeoJSON and defining default rules for areas not covered by a any polygon)

**Comment by hbruch** [comment]
> This converter works for all locations, right? I need to clarify, if we can provide other locations than stuttgart.
 
It should. But note that other areas may use different `buildSeries`.

**Review by derhuerst — APPROVED** [review]

**Pull request merged**